zenoh-ext/src/nbftreliable_subscriber.rs

```diff
@@ -0,0 +1,586 @@
+        });
+        if wait {
+            state.pending_samples.insert(source_sn, sample);
+        } else if state.last_seq_num.is_some() && source_sn != state.last_seq_num.unwrap() + 1 {
+            if source_sn > state.last_seq_num.unwrap() {
+                state.pending_samples.insert(source_sn, sample);
+            }
+        } else {
+            callback(sample);
+            let mut last_seq_num = source_sn;
+            state.last_seq_num = Some(last_seq_num);
+            while let Some(s) = state.pending_samples.remove(&(last_seq_num + 1)) {
+                callback(s);
+                last_seq_num += 1;
+                state.last_seq_num = Some(last_seq_num);
+            }
+        }
+        new
+    } else {
+        callback(sample);
+        true
+    }
+}
+
+#[zenoh_core::unstable]
+fn seq_num_range(start: Option<ZInt>, end: Option<ZInt>) -> String {
+    match (start, end) {
+        (Some(start), Some(end)) => format!("_sn={}..{}", start, end),
+        (Some(start), None) => format!("_sn={}..", start),
+        (None, Some(end)) => format!("_sn=..{}", end),
+        (None, None) => "_sn=..".to_string(),
+    }
+}
+
+#[zenoh_core::unstable]
+#[derive(Clone)]
+struct PeriodicQuery {
+    source_id: ZenohId,
+    statesref: Arc<Mutex<(HashMap<ZenohId, InnerState>, bool)>>,
+    key_expr: KeyExpr<'static>,
+    session: Arc<Session>,
+    query_target: QueryTarget,
+    query_timeout: Duration,
+    callback: Arc<dyn Fn(Sample) + Send + Sync>,
+}
+
+#[zenoh_core::unstable]
+impl PeriodicQuery {
+    fn with_source_id(mut self, source_id: ZenohId) -> Self {
+        self.source_id = source_id;
+        self
+    }
+}
+
+#[zenoh_core::unstable]
+#[async_trait]
+impl Timed for PeriodicQuery {
+    async fn run(&mut self) {
+        let mut lock = zlock!(self.statesref);
+        let (states, _wait) = &mut *lock;
+        if let Some(state) = states.get_mut(&self.source_id) {
+            state.pending_queries += 1;
+            let key_expr = (&self.source_id.into_keyexpr()) / &self.key_expr;
+            let seq_num_range = seq_num_range(Some(state.last_seq_num.unwrap() + 1), None);
+            drop(lock);
+            let handler = RepliesHandler {
+                source_id: self.source_id,
+                statesref: self.statesref.clone(),
+                callback: self.callback.clone(),
+            };
+            let _ = self
+                .session
+                .get(Selector::from(key_expr).with_parameters(&seq_num_range))
+                .callback({
+                    move |r: Reply| {
+                        if let Ok(s) = r.sample {
+                            let (ref mut states, wait) = &mut *zlock!(handler.statesref);
+                            handle_sample(states, *wait, s, &handler.callback);
+                        }
+                    }
+                })
+                .consolidation(ConsolidationMode::None)
+                .accept_replies(ReplyKeyExpr::Any)
```

Comment:
The cache replies to queries of form `<publisher_id>/k` with samples with key `k` which does not intersect. Thus the need of `Any`.

Reply:
If I got it right then, the cache declares a queryable on `<publisher_id>/k`, then it replies a stream of samples with key `k`. Then, what I believe is missing is a check that `k` actually intersects the subscription to be safe in case of a rouge queryable.

Reply:
Checks added.